src/core/lender/distributor/test/base.t.sol

```diff
@@ -0,0 +1,345 @@
+        // doesn't matter
+        senior.setReturn("balance", 200 ether);
+
+        // borrow junior: 100, senior: 50
+        balanceExpectBorrow(100 ether, 50 ether);
+
+    }
+
+    function testBorrowTakeAll() public {
+        uint amount = 300 ether;
+        shelf.setReturn("balanceRequest", requestWant, amount);
+        junior.setReturn("balance", 100 ether);
+        // doesn't matter
+        senior.setReturn("balance", 200 ether);
+
+        // borrow junior: 100, senior: 200
+        balanceExpectBorrow(100 ether, 200 ether);
+    }
+
+    function testFailBorrowTooHigh() public {
+        uint amount = 301 ether;
+        shelf.setReturn("balanceRequest", requestWant, amount);
+        junior.setReturn("balance", 100 ether);
+        // doesn't matter
+        senior.setReturn("balance", 200 ether);
+
+        // borrow junior: 100, senior: 200
+        balanceExpectBorrow(100 ether, 50 ether);
+    }
+
+    function testRepayOnlySenior() public {
+        uint amount = 50 ether;
+        shelf.setReturn("balanceRequest", !requestWant, amount);
+
+        senior.setReturn("debt", 100 ether);
+
+        distributor.balance();
+
+        assertEq(senior.calls("repay"), 1);
+        assertEq(senior.values_uint("repay_amount"), amount);
+        assertEq(senior.values_address("repay_usr"), shelf_);
+
+        assertEq(junior.calls("repay"), 0);
+    }
+
+    function testRepayOnlyJunior() public {
+        uint amount = 50 ether;
+        shelf.setReturn("balanceRequest", !requestWant, amount);
+
+        senior.setReturn("debt", 0);
+
+        distributor.balance();
+
+        assertEq(junior.calls("repay"), 1);
+        assertEq(junior.values_uint("repay_amount"), amount);
+        assertEq(junior.values_address("repay_usr"), shelf_);
+
+        assertEq(senior.calls("repay"), 0);
+    }
+
+    function testRepayOnlyJuniorWithBalance() public {
+        uint amount = 50 ether;
+        shelf.setReturn("balanceRequest", !requestWant, amount);
+
+        senior.setReturn("debt", 0);
+        junior.setReturn("balance", 100 ether);
+
+        distributor.balance();
+
+        assertEq(junior.calls("repay"), 1);
+        assertEq(junior.values_uint("repay_amount"), amount);
+        assertEq(junior.values_address("repay_usr"), shelf_);
+
+        assertEq(senior.calls("repay"), 0);
+    }
+
+    function testRepayBothTranches() public {
+        uint amount = 150 ether;
+        shelf.setReturn("balanceRequest", !requestWant, amount);
+
+        senior.setReturn("debt", 50 ether);
+
+        distributor.balance();
+
+        assertEq(senior.calls("repay"), 1);
+        assertEq(senior.values_uint("repay_amount"), 50 ether);
+        assertEq(senior.values_address("repay_usr"), shelf_);
+
+        assertEq(junior.calls("repay"), 1);
+        assertEq(junior.values_uint("repay_amount"), 100 ether);
+        assertEq(junior.values_address("repay_usr"), shelf_);
+    }
+
+    function testRepaySeniorCoveredByJuniorReserve() public {
```

Comment:
TODO please fix the repay with junior reserve logic. 
Currently the entire junior balance is transferred to the senior tranche, but it should be max the debt of the senior tranche